tests/math/test_functions.py

```diff
@@ -2266,3 +2264,138 @@ def test_sort(self, input, test_output):
+    (jnp.array([0, 0, 0, 1]), ([[0, 0], [0, 1]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1]])),
+    (torch.tensor([1, 0, 0, 0]), ([[1, 0], [0, 0]], [[1, 0], [0, 0]], [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (torch.tensor([0, 1, 0, 0]), ([[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (torch.tensor([0, 0, 1, 0]), ([[0, 0], [0, 1]], [[1, 0], [0, 0]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 0]])),
+    (torch.tensor([0, 0, 0, 1]), ([[0, 0], [0, 1]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1]])),
+    (tf.Variable([1, 0, 0, 0]), ([[1, 0], [0, 0]], [[1, 0], [0, 0]], [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (tf.Variable([0, 1, 0, 0]), ([[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (tf.Variable([0, 0, 1, 0]), ([[0, 0], [0, 1]], [[1, 0], [0, 0]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 0]])),
+    (tf.Variable([0, 0, 0, 1]), ([[0, 0], [0, 1]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1]])),
+    (tf.constant([1, 0, 0, 0]), ([[1, 0], [0, 0]], [[1, 0], [0, 0]], [[1, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (tf.constant([0, 1, 0, 0]), ([[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])),
+    (tf.constant([0, 0, 1, 0]), ([[0, 0], [0, 1]], [[1, 0], [0, 0]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 1, 0], [0, 0, 0, 0]])),
+    (tf.constant([0, 0, 0, 1]), ([[0, 0], [0, 1]], [[0, 0], [0, 1]], [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 1]])),
+]
+
+single_wires_list = [
+    ([0]),
+    ([1]),
+]
+
+multiple_wires_list = [
+    ([0, 1])
+]
+# fmt: on
+
+
+class TestDensityMatrixFromStateVectors:
+    """Tests for the density matrix for state vectors functions."""
+
+    @pytest.mark.parametrize("state_vector, expected_density_matrix", state_vectors)
+    @pytest.mark.parametrize("wires", single_wires_list)
+    def test_density_matrix_from_state_vector_single_wires(
+        self, state_vector, wires, expected_density_matrix
+    ):
+        """Test the density matrix from state vectors for single wires."""
+        density_matrix = fn.quantum._density_matrix_from_state_vector(state_vector, wires=wires)
+        assert np.allclose(density_matrix, expected_density_matrix[wires[0]])
+
+    @pytest.mark.parametrize("state_vector, expected_density_matrix", state_vectors)
+    @pytest.mark.parametrize("wires", multiple_wires_list)
+    def test_density_matrix_from_state_vector_full_wires(
+        self, state_vector, wires, expected_density_matrix
+    ):
+        """Test the density matrix from state vectors for full wires."""
+        density_matrix = fn.quantum._density_matrix_from_state_vector(state_vector, wires=wires)
+        assert np.allclose(density_matrix, expected_density_matrix[2])

```

Comment:
Just wondering if we should test more complex scenarios than these. For example, having a 3-qubit state and tracing out the second qubit.

There are some tests in `test_measurements.py` that might be good to include here.

Reply:
It is done indirectly, because using `return qml.density_matrix` will internally use this function. I might add some more complex example though 👍